micro-rdk/src/common/data_store.rs

```diff
@@ -0,0 +1,480 @@
+
+    /// Reads the next available message in the store for the given ResourceMethodKey. It should return
+    /// an empty BytesMut with 0 capacity when there are no available messages left.
+    fn read_next_message(
+        &mut self,
+        collector_key: ResourceMethodKey,
+    ) -> Result<BytesMut, DataStoreError>;
+
+    /// Resets the data store.
+    fn clear(&mut self);
+}
+
+/// StaticMemoryDataStore is an entity that governs the static bytes memory
+/// reserved in DATA_STORE. The memory is segmented based according to the DataCollectors expected
+/// (identified by collector keys) and each segment view is treated as a separate ring buffer of SensorData
+/// messages. Currently, an equal amount of space is alloted to each collector, which will affect
+/// the maximum allowed size of a single message (computed as the length of DATA_STORE divided by
+/// the number of collector keys). It should be treated as a global struct that should only be initialized once
+/// and is not thread-safe (all interactions should be blocking).
+pub struct StaticMemoryDataStore {
+    buffers: Vec<LocalRb<u8, &'static mut [MaybeUninit<u8>]>>,
+    collector_keys: Vec<ResourceMethodKey>,
+}
+
+impl StaticMemoryDataStore {
+    pub fn new(collector_keys: Vec<ResourceMethodKey>) -> Result<Self, DataStoreError> {
+        if !DATA_STORE_INITIALIZED.fetch_or(true, Ordering::SeqCst) {
+            let len_per_buffer = unsafe { DATA_STORE.len() } / collector_keys.len();
+            let mut buffers = Vec::new();
+            for i in 0..(collector_keys.len()) {
+                let start_idx = i * len_per_buffer;
+                let end_idx = (i + 1) * len_per_buffer;
+                unsafe {
+                    buffers.push(LocalRb::from_raw_parts(
+                        &mut DATA_STORE[start_idx..end_idx],
+                        0,
+                        0,
+                    ));
+                }
+            }
+            return Ok(Self {
+                buffers,
+                collector_keys,
+            });
+        }
+        Err(DataStoreError::DataStoreInitialized)
+    }
+
+    fn get_index_for_collector(
+        &self,
+        collector_key: &ResourceMethodKey,
+    ) -> Result<usize, DataStoreError> {
+        self.collector_keys
+            .iter()
+            .position(|key| key == collector_key)
+            .ok_or(DataStoreError::UnknownCollectorKey(collector_key.clone()))
+    }
+}
+
+impl DataStore for StaticMemoryDataStore {
+    fn store_message(
+        &mut self,
+        collector_key: ResourceMethodKey,
+        message: SensorData,
+    ) -> Result<(), DataStoreError> {
+        let buffer_index = self.get_index_for_collector(&collector_key)?;
+        let buffer = &mut self.buffers[buffer_index];
+        let encode_len = message.encoded_len();
+        if encode_len + 5 > buffer.vacant_len() {
+            return Err(DataStoreError::DataBufferFull(collector_key, message));
+        }
```

Comment:
Doesn't this effectively prevent overwriting old data, which I thought was the behavior we wanted? Otherwise, if we are offline and not pulling data out with `read_next_message`, we will only retain the oldest data and discard newer. Note that this will require some care, since you can't just overwrite, you need to ensure that once overwritten, that the new head points to the beginning of an encoded message.

Reply:
Wouldn't that result in dropping old data when offline even in the case where we had plenty of space to buffer instead?

Reply:
that's fair, but can't we then simply write it back to the store?

Reply:
It feels to me like there is a separation of concerns issue with that approach, because data retention / eviction logic and policy ends up mixed between the `DataStore` and the higher layer. Whereas if the `DataStore` just continually overwrites old with new when necessary, the higher layer doesn't need to know at all, I think.

Reply:
ok, so I think what I'll do in this case then is simply consume messages until the current message will fit